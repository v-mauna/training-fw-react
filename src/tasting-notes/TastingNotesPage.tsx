import React, { useEffect, useState } from 'react';
import {
  IonPage,
  IonHeader,
  IonTitle,
  IonToolbar,
  IonContent,
  IonFab,
  IonFabButton,
  IonIcon,
  IonModal,
  IonItem,
  IonLabel,
  IonList,
  IonItemOption,
  IonItemOptions,
  IonItemSliding,
} from '@ionic/react';
import { add } from 'ionicons/icons';
import TastingNoteEditor from './editor/TastingNoteEditor';
import { useTastingNotes } from './useTastingNotes';
import { TastingNote } from '../shared/models';

const TastingNotesPage: React.FC = () => {
  const [showModal, setShowModal] = useState<boolean>(false);
  const [notes, setNotes] = useState<Array<TastingNote>>([]);
  const [selectedNote, setSelectedNote] = useState<TastingNote | undefined>(
    undefined,
  );
  const { getNotes, deleteNote } = useTastingNotes();

  useEffect(() => {
    const init = async () => {
      const notes = await getNotes();
      setNotes(notes);
    };
    init();
  }, [getNotes]);

  const handleOnDismiss = async (refresh: boolean) => {
    setShowModal(false);
    setSelectedNote(undefined);
    if (refresh) {
      const notes = await getNotes();
      setNotes(notes);
    }
  };

  const handleUpdateNote = (note: TastingNote) => {
    setSelectedNote(note);
    setShowModal(true);
  };

  const handleNewNote = () => {
    setSelectedNote(undefined);
    setShowModal(true);
  };

  const handleDeleteNote = async (id: number) => {
    await deleteNote(id);
    const notes = await getNotes();
    setNotes(notes);
  };

  return (
    <IonPage>
      <IonHeader>
        <IonToolbar>
          <IonTitle>Tasting Notes</IonTitle>
        </IonToolbar>
      </IonHeader>
      <IonContent>
        <IonHeader collapse="condense">
          <IonToolbar>
            <IonTitle size="large">Tasting Notes</IonTitle>
          </IonToolbar>
        </IonHeader>
        <IonList>
          {notes.map((note, idx) => (
            <IonItemSliding key={idx}>
              <IonItem onClick={() => handleUpdateNote(note)}>
                <IonLabel>
                  <div>{note.brand}</div>
                  <div>{note.name}</div>
                </IonLabel>
              </IonItem>
              <IonItemOptions>
                <IonItemOption
                  color="danger"
                  onClick={() => {
                    handleDeleteNote(note.id!);
                  }}
                >
                  Delete
                </IonItemOption>
              </IonItemOptions>
            </IonItemSliding>
          ))}
        </IonList>
        <IonFab vertical="bottom" horizontal="end" slot="fixed">
          <IonFabButton onClick={() => handleNewNote()}>
            <IonIcon icon={add} />
          </IonFabButton>
        </IonFab>
        <IonModal isOpen={showModal}>
          <TastingNoteEditor
            onDismiss={({ refresh }) => handleOnDismiss(refresh)}
            note={selectedNote}
          />
        </IonModal>
      </IonContent>
    </IonPage>
  );
};
export default TastingNotesPage;
