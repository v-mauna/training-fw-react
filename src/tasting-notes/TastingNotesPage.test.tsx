import React from 'react';
import { render, wait, cleanup, waitForElement } from '@testing-library/react';
import { ionFireEvent as fireEvent } from '@ionic/react-test-utils';
import TastingNotesPage from './TastingNotesPage';
import { TastingNote } from '../shared/models';
jest.mock('../tea/useTea', () => ({
  useTea: () => ({
    getTeas: jest.fn(() => Promise.resolve(mockTeas)),
  }),
}));
jest.mock('./useTastingNotes', () => ({
  useTastingNotes: () => ({
    getNotes: mockGetNotes,
    getNote: mockGetNote,
  }),
}));

let mockGetNotes = jest.fn(() => Promise.resolve(mockNotes));
let mockGetNote = jest.fn(() => Promise.resolve(mockNotes[0]));

const mockNotes: Array<TastingNote> = [
  {
    id: 73,
    brand: 'Bently',
    name: 'Brown Label',
    notes: 'Essentially OK',
    rating: 3,
    teaCategoryId: 2,
  },
  {
    id: 42,
    brand: 'Lipton',
    name: 'Yellow Label',
    notes: 'Overly acidic, highly tannic flavor',
    rating: 1,
    teaCategoryId: 3,
  },
];

const mockTeas = [
  {
    id: 7,
    name: 'White',
    image: 'assets/img/white.jpg',
    description: 'White tea description.',
    rating: 5,
  },
  {
    id: 8,
    name: 'Yellow',
    image: 'assets/img/yellow.jpg',
    description: 'Yellow tea description.',
    rating: 3,
  },
];

describe('<TastingNotesPage />', () => {
  beforeEach(() => {
    mockGetNotes = jest.fn(() => Promise.resolve(mockNotes));
    mockGetNote = jest.fn(() => Promise.resolve(mockNotes[0]));
  });

  it('renders consistently', async () => {
    const { asFragment } = render(<TastingNotesPage />);
    await wait(() => expect(asFragment()).toMatchSnapshot());
  });

  describe('initialization', () => {
    it('gets all of the notes', async () => {
      render(<TastingNotesPage />);
      await wait(() => expect(mockGetNotes).toHaveBeenCalledTimes(1));
    });

    it('displays the notes', async () => {
      const { container } = render(<TastingNotesPage />);
      await wait(() => {
        expect(container).toHaveTextContent(/Bently/);
        expect(container).toHaveTextContent(/Lipton/);
      });
    });
  });

  describe('add a new note', () => {
    it('displays the editor modal', async () => {
      const { container, getByText } = render(<TastingNotesPage />);
      const button = container.querySelector('ion-fab-button')!;
      fireEvent.click(button);
      await wait(() => expect(getByText('Add New Tasting Note')).toBeDefined());
    });
  });

  describe('update an existing note', () => {
    it('prepopulates the editor modal', async () => {
      const { container, getByText } = render(<TastingNotesPage />);
      const firstNote = await waitForElement(
        () => container.querySelector('ion-item')!,
      );
      fireEvent.click(firstNote);
      await wait(() => {
        expect(getByText(/Bently/)).toBeDefined();
      });
    });
  });

  afterEach(() => {
    cleanup();
    jest.restoreAllMocks();
  });
});
