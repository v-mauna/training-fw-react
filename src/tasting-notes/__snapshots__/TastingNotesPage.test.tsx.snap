// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TastingNotesPage /> renders consistently 1`] = `
<DocumentFragment>
  <div
    class="ion-page"
  >
    <ion-header>
      <ion-toolbar>
        <ion-title>
          Tasting Notes
        </ion-title>
      </ion-toolbar>
    </ion-header>
    <ion-content>
      <ion-header
        collapse="condense"
      >
        <ion-toolbar>
          <ion-title
            size="large"
          >
            Tasting Notes
          </ion-title>
        </ion-toolbar>
      </ion-header>
      <ion-list>
        <ion-item-sliding>
          <ion-item>
            <ion-label>
              <div>
                Bently
              </div>
              <div>
                Brown Label
              </div>
            </ion-label>
          </ion-item>
          <ion-item-options>
            <ion-item-option
              color="danger"
            >
              Delete
            </ion-item-option>
          </ion-item-options>
        </ion-item-sliding>
        <ion-item-sliding>
          <ion-item>
            <ion-label>
              <div>
                Lipton
              </div>
              <div>
                Yellow Label
              </div>
            </ion-label>
          </ion-item>
          <ion-item-options>
            <ion-item-option
              color="danger"
            >
              Delete
            </ion-item-option>
          </ion-item-options>
        </ion-item-sliding>
      </ion-list>
      <ion-fab
        horizontal="end"
        slot="fixed"
        vertical="bottom"
      >
        <ion-fab-button>
          <ion-icon
            icon="data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' class='ionicon' viewBox='0 0 512 512'><path stroke-linecap='round' stroke-linejoin='round' d='M256 112v288M400 256H112' class='ionicon-fill-none ionicon-stroke-width'/></svg>"
          />
        </ion-fab-button>
      </ion-fab>
    </ion-content>
  </div>
</DocumentFragment>
`;
